File: SolarSharp/rendering/imgui/ImGizmo.cs
using SolarSharp.EngineAPI;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SolarSharp.Rendering
{
	public enum ImGizmoOperation
	{
		TRANSLATE = 0,
		ROTATE,
		SCALE,
		BOUNDS,
	};

	public enum ImGizmoMode
	{
		LOCAL = 0,
		WORLD
	};

	public class ImGizmo
    {
        const string DLLName = "SolarWindows";
        public static void Enable(bool enable) 
			=> ImGizmoAPI.GizmoEnable(enable);
        public static void SetRect(float x, float y, float width, float height) 
			=> ImGizmoAPI.GizmoSetRect(x, y, width, height);
        public static bool Manipulate(Matrix4 proj, Matrix4 view, ref Matrix4 world, ImGizmoOperation operation, ImGizmoMode mode) 
			=> ImGizmoAPI.GizmoManipulate(proj,view, ref world, (int)operation, (int)mode);
    }
}

File: SolarWindows/src/ImGizmo.cpp

#include "Core.h"
#include <memory.h>
#include "vendor/imgui/imgui.h"
#include "vendor/imgizmo/ImGuizmo.h"

#include "SolarMath.h"

EDITOR_INTERFACE(void) ImGizmoEnable(bool enable)
{
	ImGuizmo::Enable(enable);
}

EDITOR_INTERFACE(void) ImGizmoSetRect(float x, float y, float width, float height)
{
	ImGuizmo::SetRect(x, y, width, height);
}

EDITOR_INTERFACE(bool) ImGizmoIsUsing()
{
	return ImGuizmo::IsUsing();
}

EDITOR_INTERFACE(bool) ImGizmoManipulate(Mat4f proj, Mat4f view, Mat4f* world, int operation, int mode, Mat4f *deltaMatrix, float* snap)
{
	if (snap[0] == 0.0f)
		return ImGuizmo::Manipulate(view.ptr, proj.ptr, (ImGuizmo::OPERATION)operation, (ImGuizmo::MODE)mode, world->ptr, deltaMatrix->ptr);
	return ImGuizmo::Manipulate(view.ptr, proj.ptr, (ImGuizmo::OPERATION)operation, (ImGuizmo::MODE)mode, world->ptr, deltaMatrix->ptr, snap);	
}

